types/coin.go

```diff
@@ -508,15 +513,49 @@ func ParseCoin(coinStr string) (coin Coin, err error) {
 
 	denomStr, amountStr := matches[2], matches[1]
 
-	amount, ok := NewIntFromString(amountStr)
-	if !ok {
-		return Coin{}, fmt.Errorf("failed to parse coin amount: %s", amountStr)
-	}
-
 	if denomStr != strings.ToLower(denomStr) {
 		return Coin{}, fmt.Errorf("denom cannot contain upper case characters: %s", denomStr)
 	}
 
+	// test if the amount is specified using scientific notation
+	scinMatches := reScinAmt.FindStringSubmatch(amountStr)
+
+	ok := false
+	var amount Int;
+
+	if len(scinMatches) == 4 {
+
+		// Parse scientific notation such as 1.23e+12stake
+		var f float64
+		var dec int
+		
+		if f, err = strconv.ParseFloat(scinMatches[1], 64); err != nil {
+			return
+		}
+		
+		if dec, err = strconv.Atoi(scinMatches[3]); err != nil {
+			return 
+		}
+
+		// adjust dec by the length of the decimal part in the number
+		d := reDecPart.FindStringSubmatch(scinMatches[1])
+		if len(d) > 1 {
+			l := len(d[1])
+			if l > dec {
```

Comment:
If there is another way which this can be structured to avoid these tightly nested if statements we should use that (for ex. returning on `if len(d) <= 1`)